processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java

```diff
@@ -339,6 +341,33 @@ else if ( !method.isUpdateMethod() ) {
                             existingVariableNames
             );
 
+            Type actualReturnType = method.getReturnType();
+            List<LifecycleCallbackMethodReference> beforeMappingReferencesWithFinalizedReturnType = new ArrayList<>();
+            List<LifecycleCallbackMethodReference> afterMappingReferencesWithFinalizedReturnType = new ArrayList<>();
+            if ( !actualReturnType.isVoid() && !actualReturnType.equals( resultTypeToMap ) ) {
+                beforeMappingReferencesWithFinalizedReturnType.addAll( filterMappingTarget(
+                    LifecycleMethodResolver.beforeMappingMethods(
+                        method,
+                        actualReturnType,
+                        selectionParameters,
+                        ctx,
+                        existingVariableNames
+                    ),
+                    false
+                ) );
+
+                removeMappingReferencesWithoutSourceParameters(beforeMappingReferencesWithFinalizedReturnType);
+
+                afterMappingReferencesWithFinalizedReturnType.addAll( LifecycleMethodResolver.afterMappingMethods(
+                    method,
+                    actualReturnType,
+                    selectionParameters,
+                    ctx,
+                    existingVariableNames
+                ) );
+                removeMappingReferencesWithoutSourceParameters(afterMappingReferencesWithFinalizedReturnType);
+            }
+
```

Comment:
Doesn't it make more sense to initialize this in the same place where we create the `finalizeMethod` (i.e. line 401)? I think that in that case we won't need to do the if checks that are here, since we know we only need those lifecycles when using builders